sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -955,37 +955,50 @@ class GpuDynamicPartitionDataConcurrentWriter(
       subTables.head
     }
 
+    val maxRecordsPerFile = description.maxRecordsPerFile
     withResource(t) { _ =>
-      if (description.maxRecordsPerFile > 0 &&
-          status.writerStatus.recordsInFile + t.getRowCount >= description.maxRecordsPerFile) {
-        // TODO: This does not handle batch splitting to make sure maxRecordsPerFile
-        //  is not exceeded. See: https://github.com/NVIDIA/spark-rapids/issues/6452
-        // write data
-        val batch = GpuColumnVector.from(t, outDataTypes)
-        // the batch is closed in `write`
-        status.writerStatus.outputWriter.write(batch, statsTrackers)
+      val batch = GpuColumnVector.from(t, outDataTypes)
+      if (!needSplitBatch(maxRecordsPerFile, status.writerStatus.recordsInFile, batch.numRows())) {
         statsTrackers.foreach(_.newBatch(batch))
-        status.writerStatus.recordsInFile += t.getRowCount
+        status.writerStatus.recordsInFile += batch.numRows()
+        status.writerStatus.outputWriter.write(batch, statsTrackers)
```

Comment:
`batch` is not closed on this branch

Reply:
The function `write` will close the batch: